xdsl/dialects/onnx.py

```diff
@@ -0,0 +1,105 @@
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+)
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.utils.exceptions import VerifyException
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class ElementwiseBinOpBase(IRDLOperation):
+    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+    lhs = operand_def(TensorType[T])
+    rhs = operand_def(TensorType[T])
+    res = result_def(TensorType[T])
+    assembly_format = "`(` $lhs `,` $rhs `)` attr-dict `:` `(` type($lhs) `,` type($rhs) `)` `->` type($res)"
+
+    def __init__(
+        self,
+        lhs: SSAValue,
+        rhs: SSAValue,
+    ):
+        super().__init__(
+            operands=[lhs, rhs],
+            result_types=[[]],
+        )
+
+    def verify_(self) -> None:
+        # Check that the arguments are broadcastable (using Numpy semantics) and that the result type is correct.
+        res_shape: list[int] = []
+
+        # Iterate over the shapes in reverse order and compute the result shape.
+        if (
+            not isinstance(lhs_type := self.lhs.type, TensorType)
+            or not isinstance(rhs_type := self.rhs.type, TensorType)
+            or not isinstance(res_type := self.res.type, TensorType)
+        ):
+            raise ValueError(
+                "onnx elementwise binary operation operands and result must be of type TensorType"
+            )
+        lhs_shape = lhs_type.get_shape()
+        rhs_shape = rhs_type.get_shape()
+        i = max(len(lhs_shape), len(rhs_shape))
+        while i > 0:
+            i -= 1
+            d1: int = lhs_shape[i] if i >= 0 else 1
+            d2: int = rhs_shape[i] if i >= 0 else 1
+            if d1 == d2:
+                res_shape.append(d1)
+                continue
+            if d1 == 1:
+                res_shape.append(d2)
+                continue
+            if d2 == 1:
+                res_shape.append(d1)
+                continue
+            raise VerifyException(
+                f"operands have incompatible shapes: {lhs_shape} and {rhs_shape}"
+            )
+        # Reverse the result shape and check that it matches the result type.
+        res_type_shape = list(res_type.get_shape())
+        res_shape.reverse()
+        if len(res_shape) != len(res_type_shape) or res_shape != res_type_shape:
+            raise VerifyException(
+                f"result shape {res_shape} does not match result type {self.res.type}"
+            )
+
+
+@irdl_op_definition
+class Add(ElementwiseBinOpBase):
+    name = "onnx.Add"
+
+
+@irdl_op_definition
+class Sub(ElementwiseBinOpBase):
+    name = "onnx.Sub"
+
+
+@irdl_op_definition
+class Mul(ElementwiseBinOpBase):
+    name = "onnx.Mul"
+
+
+@irdl_op_definition
+class Div(ElementwiseBinOpBase):
+    name = "onnx.Div"
+
+
+ONNX = Dialect("onnx", [Add, Sub, Mul, Div])
```

Comment:
nit:
```suggestion
ONNX = Dialect("onnx", [Add, Sub, Mul, Div,])
```
Adding a comma here will make the formatter move the ops to their own lines, which will minimise diffs in the future. A really minor nit, feel free to ignore :)